### DISCUSSION TITLE
Control SELECT statement issued for ORM Eager loaded relationship

### DISCUSSION THREAD

**Discussion by Neah-Ko** [sqlalchemy/sqlalchemy#11482] [Usage Questions] [answered]
Hello,

## Context
Last time I opened a discussion it was to produce an abstract statement in order to query a table passing a list of `fields` (and `nested` representing relationship fields) as a parameter to select that looks like this:

```python
        stmt = select(table)
        stmt = stmt.options(
        load_only(
            *[getattr(table, f) for f in fields]
        ),
        *[
            selectinload(getattr(table, n))
            for n in nested
        ]
```

It works great and I can use it as a base for some filtering, add join, where and so on.

I am implementing some API that requires Group Based/R.B.A.C type of permission system for access rights.

I have a working function that takes the above statement and `join` it against a chain of multiple tables in order to dynamically check requesting user access rights  

## Issue

I would like to be able to apply those permissions for relationships as well.
That may likely have more restrictive permissions access.

For my statement building that means that I would like to be able to plug in another `select` statement which would be produced with the same function but for the `target` table (and restricted on non relationship fields as depth 1 is enough and to avoid recursion issues). 

However, I am unable to do so. 

### Subquery

I tried using `subquery()` on the nested attribute statement, using relationship `local_remote_pairs` to join back on the table, in conjunction with `.options(contains_eager())` on the name of the relationship.

From the documentation it is a bit unclear to me whether `contains_eager` actually produces a statement or simply notifies the core that this data should be part of the query result. 
I am using it as the later and also tried without.
 
It produces a statement that is close to the truth, but unfortunately I get those errors on execution:
```
sqlalchemy.exc.ProgrammingError: (sqlalchemy.dialects.postgresql.asyncpg.ProgrammingError) <class 'asyncpg.exceptions.UndefinedTableError'>: missing FROM-clause entry for table "USER"
```

which is caused by the tables in the inner select statements not being seen from the outer select statement querying the given fields.
Subqueries are then aliased as `anon_1`, `anon_2`... 

Should I pass all relationship tables alongside top table in my select statement ? Doesn't that means that data will get loaded twice ?

### Eager Loading

If I understand correctly all the functions dedicated to ORM Eager loading such as `joinload`, `selectinload`, `lazyload` and co.  are part of the > `Load Interface`  which you cannot join

To fetch nested collection data docs are promoting chain Loading into filtering.
My prototypes ideas felt a bit weird as I struggle a bit with the flat structure of the queries. 
Also each table has to fetch its permission on a variable amount of table following one-to-many relationships. 

However, would that be on the principle strictly equivalent to using select table and join ?  

Overall, would there be a way to control the select statement emitted by Load other than directly configuring relationship's `primaryjoins`/`secondaryjoin` ? 

Best,

**Comment by zzzeek** [comment]
hi -

based on your use case, rather than writing select statements why not use `selectinload(Entity.relationship).load_only('a', 'b', 'c')` ?  `load_only` can be chained onto relationship loaders.

**Reply by Neah-Ko** [reply]
Hello,

when testing loading the relationship that way I get this error:
> AttributeError: Neither 'InstrumentedAttribute' object nor 'Comparator' object associated with Entity.relationship has an attribute 'load_only'

However, I may have open this issue to quick and I apologize about it. 
I managed to get expected result, reworking my `subquery` solution with the following elements:
- nothing changes for regular fields
- each request relationship attribute is joined explicitly onto top level statement 
- for each permission attached on that relationship target table, the collection filtering statement is produced in a `subquery` and joined on statement setting filter rules for that attribute
  -  which are the same as if querying the table directly
  
In other words, what I was looking for is a way to emit a custom select statement to load an ORM relationship collection, which I managed this way:
```python
stmt # <- top level select(table) statement with my options

# For one relationship:
stmt = stmt.join(getattr(table, relationship.key)).join(
     select(relationship.target_table)
     .join(x)
     ...
     .join(y)
     ...
    .where(z)
    .subquery() # Important, else SQLA >= 2.0 raises an error.
)
```
where `relationship.target_table` is the declarative class for `relationship.target`

This may not be fully optimal as identical tables may be loaded from the different subqueries.
However, else I run into some 'table declared more than once' errors which is tackled by doing it in a subquery.

So I have solved my issue and thank you for your time.

Just out of curiosity, do you think that If I attempt avoiding subqueries and join everything _flat_ on the top level query using `aliased` tables instead I would get better performances ?

**Reply by zzzeek** [reply]
OK, just to note, what I meant was, referring to your original example

```py
        stmt = select(table)
        stmt = stmt.options(
        load_only(
            *[getattr(table, f) for f in fields]
        ),
        *[
            selectinload(getattr(table, n)).load_only(*[getattr(related, f) for f in related_fields])
            for n in nested
        ]
```


if you want to write custom selects that you invoke directly, and then wish to populate relationships with that, you need to write that out, making use of [set_committed_value](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.attributes.set_committed_value).  for example here's how to write your own selectinload recipe:  https://github.com/sqlalchemy/sqlalchemy/wiki/DisjointEagerLoading

**Comment by CaselIT** [comment, marked as answer]
Hi,

If you want to add custom criteria while loading the queries you can do so also by adding https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#sqlalchemy.orm.with_loader_criteria

**Reply by Neah-Ko** [reply]
Oh I see, that seem to be the right tool in the box actually ^^'.
I'll probably have to re-implement my current solution using this.

Thank you so much, and apologies I had missed that quite clear bit of the documentation.
